src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs

```diff
@@ -56,7 +56,7 @@ private void Error(ErrorCode code, BoundNode node, params object[] args)
 
         private void CheckUnsafeType(BoundExpression e)
         {
-            if (e != null && (object)e.Type != null && e.Type.TypeKind == TypeKind.Pointer) NoteUnsafe(e);
+            if (e != null && (object)e.Type != null && e.Type.IsPointerOrFunctionPointer()) NoteUnsafe(e);
```

Comment:
>e.Type.IsPointerOrFunctionPointer()) [](start = 55, length = 36)

Should we use ```IsUnsafe``` extension method here instead?

Reply:
No. `IsUnsafe` will check inside array types, which this is specifically not doing. For example, this is valid C# today, and would be invalid if we used `IsUnsafe` here:
```cs
using System;
using System.Linq.Expressions;

public class C {
    public unsafe void M(int*[] ptr1, int*[] ptr2) {
        Expression<Func<bool>> expr = () => ptr1 == ptr2;
    }
}
```

---
In reply to: [407750632](https://github.com/dotnet/roslyn/pull/43248#discussion_r407750632) [](ancestors = 407750632)

Reply:
FWIW, `IsUnsafe` is also intentionally incorrect to match native compiler behavior, as it doesn't recursively check generic type parameters (matching Dev10 behavior, according to the comment.)

---
In reply to: [407820700](https://github.com/dotnet/roslyn/pull/43248#discussion_r407820700) [](ancestors = 407820700,407750632)